### ISSUE TITLE
Wrong cast

### ISSUE THREAD

**Issue by ElvisBlue** [Vector35/binaryninja-api#4800]
**Version and Platform (required):**
- Binary Ninja Version: 3.6.4706-dev Personal
- OS: Windows
- OS Version: Windows 11
- CPU Architecture: x64

**Bug Description:**
Wrong cast

**Steps To Reproduce:**
Please provide all steps required to reproduce the behavior:
1. Open "challenge" file
2. Goto 0xdbf
3. Change windows to Pseudo C

**Expected Behavior:**
read_long() should be casted to (int32_t) instead of (int64_t)

**Screenshots:**
![image](https://github.com/Vector35/binaryninja-api/assets/32590178/3dd31223-a43b-49d7-a8b9-6bb777adf04f)

**Comment by fuzyll** [comment]
Am I missing something? This _seems_ correct to me:

<img width="875" alt="image" src="https://github.com/Vector35/binaryninja-api/assets/607452/400b1300-0c1a-4afc-bd60-9193942df168">

After the `call` to `read_long` (which is typed as returning a `uint64_t`), there is a `cdqe` instruction that will sign-extend `eax`. This is why the cast is shown: It's casting an unsigned value to a signed value. It's possible we got the type of `read_long` incorrect (its return value appears to come from a `strtoul`, which we've typed as `uint64_t`) because that `cdqe` suggests it should be 32-bit. But, regardless, I'm pretty sure `rax_8` is the value _after_ the `cdqe`, which really is `int64_t`.

If I'm wrong or if I've missed something important in my reading of what's going on, please let me know. Otherwise, I _believe_ we're doing the right things here.

**Comment by ElvisBlue** [comment]
Hi @fuzyll 
Sorry for that I am not native English speaker so explaining something technical in English is bit hard. But both IDA and ghidra cast it correctly

![image](https://github.com/Vector35/binaryninja-api/assets/32590178/84a7e729-2eea-483b-933d-a5639babc63b)

![image](https://github.com/Vector35/binaryninja-api/assets/32590178/748a2163-3598-4c8e-bb6a-c5752a0cd801)

Also here is well explaination about `cdeq` instruction I found on Internet
https://stackoverflow.com/questions/54618685/what-is-the-meaning-use-of-the-movzx-cdqe-instructions-in-this-code-output-by-a

**Comment by fuzyll** [comment]
No problem! I think your English is good enough. 🙂 

Your second screenshot shows Ghidra casting the value to a `long`, which is (as far as I know) equivalent to `int64_t`. So, as far as I can tell, it's just IDA that thinks it's a 32-bit value.

What do Ghidra and IDA say the return type of the `read_long` function is? Binary Ninja thinks it's `uint64_t`. If you change it to `uint32_t`, the cast (and type of `rax_8`) will change, too, if that's what you want.

**Comment by ElvisBlue** [comment]
I think ghidra cast it correctly. `read_long` returns uint64_t type, after `cdeq` instruction, the returned value cast to int32_t before assigned to 64 bit variable. That why ghidra show `(long)(int)uVar1` instead of `(long)uVar1`. What I mean here is binary ninja should cast it to 32 bit type, then assign to 64 bit variable instead of cast it to 64 bit type then assign to 64 bit variable.

**Comment by ElvisBlue** [comment]
Hi @fuzyll any progress on this? Or still you think this is intended?

**Comment by fuzyll** [comment]
Hey, sorry, this was about a week before I went on my holiday break and I must've missed responding earlier.

So, looking at this further, I see your point: Because the return value from `read_long` is treated as a 32-bit value (because we're doing `cdqe`), we should have two casts to show that the value is being truncated and then extended.

In HLIL, at that location, we emit `int64_t rax_8 = sx.q(read_long())`, which correctly understands that we have `eax` and are setting `rax` due to the `cdqe`:

![image](https://github.com/Vector35/binaryninja-api/assets/607452/040c1cec-5ca5-4f84-9383-a592984230b0)

This is actually a little confusing because `read_long` returns a 64-bit value. According to [our own documentation](https://docs.binary.ninja/dev/bnil-overview.html#example), I would expect us to show a `.d` somewhere (like `read_long().d` maybe?). We don't, but it does appear we get the rest of this right.

So, I would guess the problem here is that we don't emit some necessary token in this case that shows the `.d` (in HLIL) or `(int32_t)` (in Pseudo-C.

TL;DR: I agree this is an issue. Just not sure what to do with it yet. Thanks for poking me to look at it again.

**Comment by fuzyll** [comment]
Rusty thinks this is possibly due to the partial access resolver and is very concerned that a fix here might have big problems for floating point support. Not entirely sure what the right path forward is, but I've done my best to complete triage of the issue now.

**Comment by ElvisBlue** [comment]
Ahhhh thank you for not forgeting this

**Comment by xusheng6** [comment]
This looks similar, but not necessarily related to https://github.com/Vector35/binaryninja-api/issues/6595
